File: src/engine/risk.ts
import type { Order, RiskLimits } from "../types";

export interface RiskContext {
  equity: number;
  dayStartEquity: number;
  realizedPnl: number;
  unrealizedPnl: number;
  grossExposure: number; // absolute notional across all positions
  positionNotional: (marketId: string) => number;
  ordersLastMin: (strategyId: string) => number;
  strategyExposure: (strategyId: string) => number;
  dataAgeSec: (marketId: string) => number;
}

export interface RiskDecision {
  approved: boolean;
  reason?: string;
  /** possibly-reduced quantity after sizing caps */
  qty: number;
}

export const DEFAULT_LIMITS: RiskLimits = {
  killSwitch: false,
  maxDailyLossPct: 5,
  maxPositionPct: 15,
  maxGrossExposurePct: 60,
  perStrategyBudgetPct: 20,
  kellyFraction: 0.25,
  maxOrdersPerMin: 10,
  maxDataStalenessSec: 30,
  maxDrawdownPct: 15,
  stopAtrMult: 3,
  takeProfitAtrMult: 5,
  trailingAtrMult: 0,
  maxConsecutiveLosses: 4,
  cooldownSec: 300,
  volTargetPct: 0,
  regimeFilter: false,
};

// The risk manager is sovereign: this is the ONLY path from a strategy intent to
// a routed order. It fails closed — any doubt rejects the order.
export function evaluate(
  order: Order,
  price: number,
  limits: RiskLimits,
  ctx: RiskContext
): RiskDecision {
  const notional = order.qty * price;

  // 1. Kill switch — only closing/flatten intents pass when tripped.
  if (limits.killSwitch && order.side === "buy") {
    return { approved: false, qty: 0, reason: "kill switch active" };
  }

  // 2. Stale data gate.
  const age = ctx.dataAgeSec(order.marketId);
  if (age > limits.maxDataStalenessSec) {
    return { approved: false, qty: 0, reason: `stale data (${age.toFixed(0)}s)` };
  }

  // 3. Daily loss ceiling.
  const dayPnl = ctx.realizedPnl + ctx.unrealizedPnl;
  const dayLossPct = (-dayPnl / ctx.dayStartEquity) * 100;
  if (dayLossPct >= limits.maxDailyLossPct) {
    return {
      approved: false,
      qty: 0,
      reason: `daily loss limit hit (${dayLossPct.toFixed(1)}%)`,
    };
  }

  // 4. Rate limit per strategy.
  if (ctx.ordersLastMin(order.strategyId) >= limits.maxOrdersPerMin) {
    return { approved: false, qty: 0, reason: "order rate limit" };
  }

  // 5. Per-strategy budget.
  const stratCap = (limits.perStrategyBudgetPct / 100) * ctx.equity;
  if (ctx.strategyExposure(order.strategyId) + notional > stratCap) {
    return { approved: false, qty: 0, reason: "strategy budget exceeded" };
  }

  // 6. Per-position cap (may reduce qty rather than reject on buys).
  const posCap = (limits.maxPositionPct / 100) * ctx.equity;
  const curPos = Math.abs(ctx.positionNotional(order.marketId));
  let qty = order.qty;
  if (order.side === "buy" && curPos + notional > posCap) {
    const room = Math.max(0, posCap - curPos);
    if (room <= 0) {
      return { approved: false, qty: 0, reason: "position size cap" };
    }
    qty = room / price;
  }

  // 7. Gross exposure ceiling.
  const grossCap = (limits.maxGrossExposurePct / 100) * ctx.equity;
  if (order.side === "buy" && ctx.grossExposure + qty * price > grossCap) {
    const room = Math.max(0, grossCap - ctx.grossExposure);
    if (room <= 0) {
      return { approved: false, qty: 0, reason: "gross exposure cap" };
    }
    qty = Math.min(qty, room / price);
  }

  if (qty <= 0) return { approved: false, qty: 0, reason: "sized to zero" };
  return { approved: true, qty };
}

/** Fractional-Kelly position sizing, bounded to a sane fraction of equity. */
export function kellySize(
  edge: number, // p_win - p_lose style edge, 0..1
  price: number,
  equity: number,
  limits: RiskLimits
): number {
  const f = Math.max(0, Math.min(1, edge)) * limits.kellyFraction;
  const dollars = f * equity;
  return dollars / price;
}

File: crates/pythia-core/src/engine/risk.rs
//! The sovereign risk manager. Every order — paper or live — passes through
//! `evaluate` before routing. It fails closed: any doubt rejects.

use super::{RiskDecision, RiskLimits};
use crate::connectors::{OrderRequest, Side};

/// Live state the risk check needs, supplied by the portfolio ledger.
pub struct RiskContext {
    pub equity: f64,
    pub day_start_equity: f64,
    pub realized_pnl: f64,
    pub unrealized_pnl: f64,
    pub gross_exposure: f64,
    pub position_notional: f64, // for this market
    pub strategy_exposure: f64, // for this strategy
    pub orders_last_min: u32,
    pub data_age_sec: u64,
}

pub fn evaluate(
    order: &OrderRequest,
    price: f64,
    limits: &RiskLimits,
    ctx: &RiskContext,
) -> RiskDecision {
    let reject = |r: &str| RiskDecision { approved: false, qty: 0.0, reason: Some(r.into()) };
    let notional = order.qty * price;

    // 1. Kill switch — only closing (sell) intents pass when tripped.
    if limits.kill_switch && order.side == Side::Buy {
        return reject("kill switch active");
    }
    // 2. Stale data.
    if ctx.data_age_sec > limits.max_data_staleness_sec {
        return reject("stale market data");
    }
    // 3. Daily loss ceiling.
    let day_pnl = ctx.realized_pnl + ctx.unrealized_pnl;
    let day_loss_pct = (-day_pnl / ctx.day_start_equity) * 100.0;
    if day_loss_pct >= limits.max_daily_loss_pct {
        return reject("daily loss limit hit");
    }
    // 4. Rate limit.
    if ctx.orders_last_min >= limits.max_orders_per_min {
        return reject("order rate limit");
    }
    // 5. Per-strategy budget.
    let strat_cap = (limits.per_strategy_budget_pct / 100.0) * ctx.equity;
    if ctx.strategy_exposure + notional > strat_cap {
        return reject("strategy budget exceeded");
    }

    // 6. Per-position cap — reduce qty rather than reject on buys.
    let mut qty = order.qty;
    let pos_cap = (limits.max_position_pct / 100.0) * ctx.equity;
    if order.side == Side::Buy && ctx.position_notional.abs() + notional > pos_cap {
        let room = (pos_cap - ctx.position_notional.abs()).max(0.0);
        if room <= 0.0 {
            return reject("position size cap");
        }
        qty = room / price;
    }
    // 7. Gross exposure ceiling.
    let gross_cap = (limits.max_gross_exposure_pct / 100.0) * ctx.equity;
    if order.side == Side::Buy && ctx.gross_exposure + qty * price > gross_cap {
        let room = (gross_cap - ctx.gross_exposure).max(0.0);
        if room <= 0.0 {
            return reject("gross exposure cap");
        }
        qty = qty.min(room / price);
    }

    if qty <= 0.0 {
        return reject("sized to zero");
    }
    RiskDecision { approved: true, qty, reason: None }
}

/// Fractional-Kelly sizing, bounded by the caps in `evaluate`.
pub fn kelly_size(edge: f64, price: f64, equity: f64, limits: &RiskLimits) -> f64 {
    let f = edge.clamp(0.0, 1.0) * limits.kelly_fraction;
    (f * equity) / price
}
